### PR TITLE
python312: fix build on darwin

### LINKED ISSUE #208693: python3-3.12.0a3: build broken (at least on macOS)
### Steps To Reproduce
1. build python3

### Build log
```
building '/nix/store/vfqf5hykv225zvb0yba438xf6pjfg5y1-python3-3.12.0a3.drv'...
unpacking sources
unpacking source archive /nix/store/v0xdx4wij1zly1lwff89plnwknbk0cir-Python-3.12.0a3.tar.xz
source root is Python-3.12.0a3
setting SOURCE_DATE_EPOCH to timestamp 1670351720 of file Python-3.12.0a3/Misc/NEWS
patching sources
applying patch /nix/store/jqin2kzvzrvi30cxda5zp7qz1fanmz7v-no-ldconfig.patch
patching file Lib/ctypes/util.py
applying patch /nix/store/r112dk8w7zvdjipki58ch00m825li7fq-virtualenv-permissions.patch
patching file Lib/venv/__init__.py
Hunk #1 succeeded at 450 with fuzz 2 (offset 71 lines).
applying patch /nix/store/lpfs02z2i2hamz9f50p4bz6v0fs0600j-mimetypes.patch
patching file Lib/mimetypes.py
Hunk #1 succeeded at 46 (offset 6 lines).
applying patch /nix/store/qy6yc7y93ljb5ad8a0f8q5nv652lycah-darwin-libutil.patch
patching file Modules/posixmodule.c
Hunk #1 succeeded at 7263 (offset 5 lines).
applying patch /nix/store/jz4vxbi3acip31frway73anbqzkf6cwy-python-3.x-distutils-C++.patch
patching file Lib/_osx_support.py
can't find file to patch at input line 28
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/Lib/distutils/cygwinccompiler.py b/Lib/distutils/cygwinccompiler.py
|index 66c12dd358..dddb9fd2d4 100644
|--- a/Lib/distutils/cygwinccompiler.py
|+++ b/Lib/distutils/cygwinccompiler.py
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
4 out of 4 hunks ignored
can't find file to patch at input line 88
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/Lib/distutils/sysconfig.py b/Lib/distutils/sysconfig.py
|index 3414a761e7..f1af560cc1 100644
|--- a/Lib/distutils/sysconfig.py
|+++ b/Lib/distutils/sysconfig.py
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
3 out of 3 hunks ignored
can't find file to patch at input line 155
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/Lib/distutils/unixccompiler.py b/Lib/distutils/unixccompiler.py
|index d00c48981e..4a3d271fee 100644
|--- a/Lib/distutils/unixccompiler.py
|+++ b/Lib/distutils/unixccompiler.py
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
3 out of 3 hunks ignored
patching file Makefile.pre.in
Hunk #1 FAILED at 732.
1 out of 1 hunk FAILED -- saving rejects to file Makefile.pre.in.rej
error: builder for '/nix/store/vfqf5hykv225zvb0yba438xf6pjfg5y1-python3-3.12.0a3.drv' failed with exit code 1
```

### Notify maintainers
@fridh

### Metadata
```
 - system: `"aarch64-darwin"`
 - host os: `Darwin 22.2.0, macOS 13.1`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.12.0`
 - channels: `"nixpkgs, stable-21.11-darwin"`
 - nixpkgs: `~/.nix-defexpr/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

Closes #208693.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wegank** [comment]
@ofborg build python312

**Review by marsam — APPROVED** [review]

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 1]
```diff
@@ -260,7 +260,7 @@ in with passthru; stdenv.mkDerivation {
   ] ++ optionals (pythonAtLeast "3.9" && pythonOlder "3.11" && stdenv.isDarwin) [
     # Stop checking for TCL/TK in global macOS locations
     ./3.9/darwin-tcl-tk.patch
-  ] ++ optionals (isPy3k && hasDistutilsCxxPatch) [
+  ] ++ optionals (isPy3k && hasDistutilsCxxPatch && pythonOlder "3.12") [
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/python/cpython/default.nix:263]: Is now linux broken or is this unrelated?

```
building '/nix/store/ynv4z5p29s6zs84l6w6g0232zq42jiy0-python3-3.12.0a3-env.drv'...
created 252 symlinks in user environment
building '/nix/store/7vczn2znq9b9bkrv97cbsdr75zn64dr9-python3-3.12.0a3-site-prefix-mypy-test.drv'...
Traceback (most recent call last):
  File "/nix/store/f6k6kv22np8n8x71k6fbzpmnc7dlk6x4-python3.12-mypy-0.991/bin/.mypy-wrapped", line 6, in <module>
    from mypy.__main__ import console_entry
  File "/nix/store/y3xpjxh3n1apqzppr82vmabb1sv8sigj-python3-3.12.0a3-env/lib/python3.12/site-packages/mypy/__main__.py", line 9, in <module>
    from mypy.main import main, process_options
  File "mypy/main.py", line 215, in <module>
TypeError: type AugmentedHelpFormatter has the Py_TPFLAGS_MANAGED_DICT flag but tp_dictoffset is set
error: 1 dependencies of derivation '/nix/store/1xrgh2nnmxhm6sp3i5ashr9456ckbrcc-python3-3.12.0a3-env.drv' failed to build
error: builder for '/nix/store/7vczn2znq9b9bkrv97cbsdr75zn64dr9-python3-3.12.0a3-site-prefix-mypy-test.drv' failed with exit code 1;
       last 7 log lines:
       > Traceback (most recent call last):
       >   File "/nix/store/f6k6kv22np8n8x71k6fbzpmnc7dlk6x4-python3.12-mypy-0.991/bin/.mypy-wrapped", line 6, in <module>
       >     from mypy.__main__ import console_entry
       >   File "/nix/store/y3xpjxh3n1apqzppr82vmabb1sv8sigj-python3-3.12.0a3-env/lib/python3.12/site-packages/mypy/__main__.py", line 9, in <module>
       >     from mypy.main import main, process_options
       >   File "mypy/main.py", line 215, in <module>
       > TypeError: type AugmentedHelpFormatter has the Py_TPFLAGS_MANAGED_DICT flag but tp_dictoffset is set
       For full logs, run 'nix log /nix/store/7vczn2znq9b9bkrv97cbsdr75zn64dr9-python3-3.12.0a3-site-prefix-mypy-test.drv'.
error: 1 dependencies of derivation '/nix/store/k9vvfa5wfl2ha29rf2ii9i9a3a4yrs8x-python3-3.12.0a3-tests-nixenv-virtualenv.drv' failed to build
error: 1 dependencies of derivation '/nix/store/yxlxiqxprgaglk0ab7r0gwyzdlzn64nm-import-requests.drv' failed to build
error: build of '/nix/store/1xrgh2nnmxhm6sp3i5ashr9456ckbrcc-python3-3.12.0a3-env.drv', '/nix/store/7vczn2znq9b9bkrv97cbsdr75zn64dr9-python3-3.12.0a3-site-prefix-mypy-test.drv', '/nix/store/k9vvfa5wfl2ha29rf2ii9i9a3a4yrs8x-python3-3.12.0a3-tests-nixenv-virtualenv.drv', '/nix/store/yxlxiqxprgaglk0ab7r0gwyzdlzn64nm-import-requests.drv' failed
```

**Comment by wegank** [comment]
@ofborg build python312 python312.passthru.tests

File: pkgs/development/interpreters/python/cpython/default.nix (hunk 1 above)
Comment (wegank) [reply pkgs/development/interpreters/python/cpython/default.nix:263]: `hasDistutilsCxxPatch = !(stdenv.cc.isGNU or false)` evaluates to false on Linux, so the failure cannot be due to this PR.

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**